Service Attachment Creations and Update

swetharepakula opened the pull request:
ProcessServiceAttachment handles service attachment creations and updates.

swetharepakula commented:
/assign @freehan

bowei commented:
/ok-to-test

freehan commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,330 @@
+// provided connPreference
+func getConnectionPreference(connPreference string) (string, error) {
+	switch strings.TrimSpace(connPreference) {
+	case "acceptAutomatic":
```
let us keep it ver·ba·tim

swetharepakula commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,330 @@
+	frName, ok := svc.Annotations[annotations.TCPForwardingRuleKey]
+	if !ok {
+		if frName, ok = svc.Annotations[annotations.UDPForwardingRuleKey]; !ok {
+			frName = cloudprovider.DefaultLoadBalancerName(svc)
```
add a comment about the decisions here

freehan replied:
add a log line and comment herer

freehan commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,330 @@
+	foundMatchingIP := false
+	for _, ing := range svc.Status.LoadBalancer.Ingress {
+		if ing.IP == fwdRule.IPAddress {
+			foundMatchingIP = true
```
add a log line here

freehan commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,330 @@
+	}
+
+	foundMatchingIP := false
+	for _, ing := range svc.Status.LoadBalancer.Ingress {
```
double check for nil

swetharepakula commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,330 @@
+	}
+
+	foundMatchingIP := false
+	for _, ing := range svc.Status.LoadBalancer.Ingress {
```
verify that LoadBalancer is not nil

swetharepakula replied:
Verified LoadBalancer will not be empty not nil when not populated: https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3991.

freehan commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,330 @@
+
+	var subnetURLs []string
+	for _, subnetName := range subnets {
+		subnet, err := c.cloud.Compute().AlphaSubnetworks().Get(context2.Background(), meta.RegionalKey(subnetName, c.cloud.Region()))
```
Use GA Subnetwork calls

swetharepakula replied:
Thanks for catching that!

swetharepakula commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,330 @@
+	for _, subnetName := range subnets {
+		subnet, err := c.cloud.Compute().AlphaSubnetworks().Get(context2.Background(), meta.RegionalKey(subnetName, c.cloud.Region()))
+		if err != nil {
+			return subnetURLs, fmt.Errorf("failed to get Subnetwork %s: %q", subnetName, err)
```
make error more clear. add region information too

freehan replied:
add region info in the subnet name

swetharepakula commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,330 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
copyright should be 2021

freehan commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,337 @@
+
+	if existingSA != nil {
+		klog.V(4).Infof("Found existing service attachment %s", existingSA.Name)
+		err = validateUpdate(existingSA, gceSvcAttachment)
```
I am wondering if it is worth validating update. If we just let it go to call GCE API, GCE will return validation error, correct? What would be the difference?

swetharepakula replied:
That is true. I think we can do that in the future when there is an Update API. Currently since there is none in the Alpha API, I think we have to do the validation. Really it is to inform the user whatever they changed will not reflected.

freehan commented on pkg/psc/controller.go:
```diff
@@ -0,0 +1,337 @@
+	klog.V(2).Infof("Processing Service attachment %s/%s", namespace, name)
+
+	svcAttachment := obj.(*sav1alpha1.ServiceAttachment)
+	connPref, err := getConnectionPreference(svcAttachment.Spec.ConnectionPreference)
```
I am wondering if we need to validate here. Can we just pass it verbatim and let GCE API validates it.

freehan replied:
If a new type is introduced, we do not need to change code and release it to support it.

swetharepakula replied:
that's a good point, i agree, we can remove this validation.

swetharepakula commented:
Updated the PR so now we don't validate the connection preference. I updated the resource reference validation as well. Since the `Service` is in the `core` api group, the group has to either by empty or nil.

swetharepakula commented:
One additional update was made to service attachment informer creation in `context.go`.

freehan commented:
/lgtm
/approve
